### PR TITLE
nixos/maddy: Add tls option

### PR DESCRIPTION
**Pull request by onny**
###### Description of changes

Add option `tls` for maddy mail server

```
services.maddy = {
  enable = true;
  hostname = "localhost";
  ensureAccounts = [
    "admin@localhost"
    "user1@localhost"
    "user2@localhost"
  ];
  ensureCredentials = {
    "admin@localhost".passwordFile = /secrets/admin-localhost;
    "user1@localhost".passwordFile = /secrets/user1-localhost;
    "user2@localhost".passwordFile = /secrets/user2-localhost;
  };
  tls = {
    loader = "file";
    certificates = [{
      keyPath = ./localhost.key;
      certPath = ./localhost.crt;
    }];
  };
};
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/mail/maddy.nix [hunk 1]
```diff
@@ -35,6 +33,7 @@ let
       }
       optional_step file /etc/maddy/aliases
     }
+
```
Comment (RaitoBezarius) [inline nixos/modules/services/mail/maddy.nix:36]: unrelated change

File: nixos/modules/services/mail/maddy.nix [hunk 2]
```diff
@@ -215,6 +214,41 @@ in {
         '';
       };
 
+      tls = {
+
```
Comment (RaitoBezarius) [inline nixos/modules/services/mail/maddy.nix:218]: extra newline

File: nixos/modules/services/mail/maddy.nix [hunk 3]
```diff
@@ -215,6 +214,41 @@ in {
...
+          description = lib.mdDoc ''
+            TLS certificates are obtained by modules called "certificate
+            loaders". Currently only the file loader is supported which reads
+            certificates from files specifying the options `keyPaths` and
+            `certPaths`.
+          '';
+        };
+
+        keyPaths = mkOption {
+          type = types.nullOr (types.listOf types.path );
```
Comment (RaitoBezarius) [inline nixos/modules/services/mail/maddy.nix:231]: extra whitespace before `)`

File: nixos/modules/services/mail/maddy.nix [hunk 4]
```diff
@@ -215,6 +214,41 @@ in {
...
+          example = [ "/etc/ssl/mx1.example.org.key" ];
+          description = lib.mdDoc ''
+            List of paths to the private keys used for TLS. Maddy will use SNI
+            if multiple key and certificate pairs are selected.
+          '';
+          default = null;
+        };
+
+        certPaths = mkOption {
+          type = types.nullOr (types.listOf types.path );
```
Comment (RaitoBezarius) [inline nixos/modules/services/mail/maddy.nix:241]: extra whitespace before `)`

**Review by RaitoBezarius — APPROVED** [review]
LGTM modulo formatting; I don't use maddy; untested

File: nixos/modules/services/mail/maddy.nix [hunk 5]
```diff
@@ -215,6 +214,39 @@ in {
...
+          default = "off";
+          description = lib.mdDoc ''
+            TLS certificates are obtained by modules called "certificate
+            loaders". Currently only the file loader is supported which reads
+            certificates from files specifying the options `keyPaths` and
+            `certPaths`.
+          '';
+        };
+
+        keyPaths = mkOption {
```
Comment (e1mo) [inline nixos/modules/services/mail/maddy.nix:229]: Why don't we go with a sub module in the style of `services.madd.tls.certificates = [{ key = "/var/lib/ssl/mydomain.key"; cert = "/var/lib/ssl/mydomain.cert"; }]`here?

I don't exactly know how / if list ordering is handles / preserved when merging multiple definitions, but I think having two lists whose values are only used linked with each other, is more fragile and less intuitive compared to `listOf submodule`.

File: nixos/modules/services/mail/maddy.nix [hunk 6]
```diff
@@ -215,6 +214,39 @@ in {
         '';
       };
 
+      tls = {
+        loader = mkOption {
+          type = types.enum [ "file" "off" ];
```
Comment (e1mo) [inline nixos/modules/services/mail/maddy.nix:219]: What does limit our ability to implement the other loader? Is it only limited by our ability to include the domain registrar tokens from a file?

**Review by e1mo — CHANGES_REQUESTED** [review]
So far, I have only done a nix code review and not tested it (neither do I run maddy). And while to code itself seems fine, I think the configuration style is the best of what would be possible in NixOS modules.

Further more, I'd kinda like to see a integration with `security.acme` so that users will only need to pass the acmeDomains whose certificates need to be included and let the module handle permissions, service dependencies and so on. A full-fledged `security.acme` integration (aka. `enableAme` to auto request and use certificates instead of having the user manually specify them manually) would also be neat, but that is probably wishful thinking.

The non-removable `tls off` is the biggest active merge blocker for me.

File: nixos/modules/services/mail/maddy.nix (hunk 6 above)
Comment (onny) [reply nixos/modules/services/mail/maddy.nix:219]: I would like to stick with only implementing `file` loader in this PR since it will be the only one I'm going to use

[the remark this answers is not in the document]
Comment (onny) [reply nixos/modules/services/mail/maddy.nix:219]: That is a good idea

File: nixos/modules/services/mail/maddy.nix [hunk 7]
```diff
@@ -318,6 +362,15 @@ in {
         $(primary_domain) = ${cfg.primaryDomain}
         $(local_domains) = ${toString cfg.localDomains}
         hostname ${cfg.hostname}
+
+        ${if (cfg.tls.loader == "file") then ''
+          tls file ${toString (
+            zipListsWith (a: b: a + " " +  b) cfg.tls.certPaths cfg.tls.keyPaths
```
[the remark this answers is not in the document]
Comment (onny) [reply nixos/modules/services/mail/maddy.nix:368]: Since I wont going to use this feature I thought about skipping it for this PR. Do you think it should be integrated?

**Comment by onny** [comment]
> Further more, I'd kinda like to see a integration with `security.acme` so that users will only need to pass the acmeDomains whose certificates need to be included and let the module handle permissions, service dependencies and so on. A full-fledged `security.acme` integration (aka. `enableAme` to auto request and use certificates instead of having the user manually specify them manually) would also be neat, but that is probably wishful thinking.

That would be neat. Could be an addition for an followup PR. Having this `file` loader might be a good starting point to further figure out how to integrate something like `enableACME`

File: nixos/modules/services/mail/maddy.nix (hunk 5 above)
Comment (e1mo) [reply nixos/modules/services/mail/maddy.nix:229]: I don't know of any explicitly using this style, but I also don't know any module doing it the proposed way I think.
All modules I think of have a 1:1 mapping of Domain / instance / … to Certificates. But since maddy doesn't do that, we sadly can't do the same.

File: nixos/modules/services/mail/maddy.nix (hunk 6 above)
Comment (e1mo) [reply nixos/modules/services/mail/maddy.nix:219]: Okay. And if the `security.acme` will ever see the light of day (I try to refrain myself from doing it for now, because I kinda should study for my A-Levels instead of doing that ;D) we would also have support for a wild variety of DNS Challenge providers.

**Comment by onny** [comment]
> On extra thing I'd just like to mention for the record, with changes like these the meaning of `services.maddy.config` is moving more and more to a meaning I'd expect from `extraConfig`. Since more and more settings are configured in a Non-overridable manner based on the settings, while `config` suggest full configurability. This is currently only applicable to a very limited number of settings, so I'm fine with it. But if more and more settings are moved to structured settings (which I'm fully in favor of), I believe `config` will need to be renamed to `extraConfig` at some point (or also implement the structured and literal config dualism known from nginx).

Good point. This is a change I could do when something like a `maddy.listeners = []` option is implemented, which would configure smtp and imap sockets.

**Comment by onny** [comment]
Thank you @e1mo for reviewing. Integrated the changes :)

**Comment by onny** [comment]
I updated the assertion to the following format:
```
    assertions = [{
      assertion = (cfg.tls.loader == "file") -> (
        cfg.tls.certificates != null &&
        length cfg.tls.certificates > 0 &&
        all (attrs: attrs ? certPath && attrs ? keyPath) cfg.tls.certificates
      );
      message = ''
        If maddy is configured to use TLS, tls.certificates with attribute
        pairs of certPath and keyPath must be provided.
        Read more about obtaining TLS certificates here:
        https://maddy.email/tutorials/setting-up/#tls-certificates
      '';
    }];
```
It does check if entries are present in the `certificates` list but unfortunately the last check if both keys are present in the attribute set doesn't seem to work, it still tries to evaluate and fails:
```
error: A definition for option `services.maddy.tls.certificates."[definition 1-entry 2]".certPath' is not of type `path'. Definition values:
       - In `/home/onny/projects/nixpkgs/nixos/modules/services/mail/maddy.nix': null
(use '--show-trace' to show detailed location information)
```

File: nixos/modules/services/mail/maddy.nix [hunk 8]
```diff
@@ -270,6 +328,20 @@ in {
...
 
+    assertions = [{
+      assertion = cfg.tls.loader == "file" -> (
+        cfg.tls.certificates != null &&
+        length cfg.tls.certificates > 0 &&
+        all (attrs: attrs ? certPath && attrs ? keyPath) cfg.tls.certificates
+      );
+      message = ''
+        If maddy is configured to use TLS, tls.certificates with attribute
+        pairs of certPath and keyPath must be provided.
```
Comment (e1mo) [inline nixos/modules/services/mail/maddy.nix:339]:
```suggestion
        sets with certPath and keyPath must be provided.
```

File: nixos/modules/services/mail/maddy.nix [hunk 9]
```diff
@@ -270,6 +328,20 @@ in {
 
   config = mkIf cfg.enable {
 
+    assertions = [{
+      assertion = cfg.tls.loader == "file" -> (
+        cfg.tls.certificates != null &&
+        length cfg.tls.certificates > 0 &&
+        all (attrs: attrs ? certPath && attrs ? keyPath) cfg.tls.certificates
```
Comment (e1mo) [inline nixos/modules/services/mail/maddy.nix]: I'm not sure, but shouldn't the existence of these attributes be checked by the submodule? (Not 100% sure)

File: nixos/modules/services/mail/maddy.nix [hunk 10]
```diff
@@ -215,6 +214,65 @@ in {
...
+          type = with types; nullOr (listOf (submodule {
+            options = {
+              keyPath = mkOption {
+                type = with types; nullOr path;
+                example = "/etc/ssl/mx1.example.org.key";
+                default = null;
+                description = lib.mdDoc ''
+                  Path to the private key used for TLS.
+                '';
+              };
```
Comment (e1mo) [inline nixos/modules/services/mail/maddy.nix:239]: Why allow null here? If any attrset of this kind exists in the list, it requires both options to be set. I think it'd be best to represent this in the options instead of adding assertions on top.

```suggestion
                type = types.path;
                example = "/etc/ssl/mx1.example.org.key";
                description = lib.mdDoc ''
                  Path to the private key used for TLS.
                '';
              };
```

File: nixos/modules/services/mail/maddy.nix [hunk 11]
```diff
@@ -215,6 +214,65 @@ in {
...
+              certPath = mkOption {
+                type = with types; nullOr path;
+                example = "/etc/ssl/mx1.example.org.crt";
+                default = null;
+                description = lib.mdDoc ''
+                  Path to the certificate used for TLS.
+                '';
+              };
```
Comment (e1mo) [inline nixos/modules/services/mail/maddy.nix:247]:
```suggestion
              certPath = mkOption {
                type = types.path;
                example = "/etc/ssl/mx1.example.org.crt";
                description = lib.mdDoc ''
                  Path to the certificate used for TLS.
                '';
              };
```

File: nixos/modules/services/mail/maddy.nix [hunk 12]
```diff
@@ -215,6 +214,65 @@ in {
...
+            };
+          }));
+          default = null;
```
Comment (e1mo) [inline nixos/modules/services/mail/maddy.nix:250]: See above
```suggestion
          default = [];
```

File: nixos/modules/services/mail/maddy.nix [hunk 13]
```diff
@@ -215,6 +214,65 @@ in {
...
+          description = lib.mdDoc ''
+            TLS certificates are obtained by modules called "certificate
+            loaders". Currently only the file loader is supported which reads
+            certificates from files specifying the options `keyPaths` and
+            `certPaths`.
+          '';
+        };
+
+        certificates = mkOption {
+          type = with types; nullOr (listOf (submodule {
```
Comment (e1mo) [inline nixos/modules/services/mail/maddy.nix:230]: Why `null` instead of an empty list? I think an empty list is the better choice for "no certificates", I'd see `null` as a "certificates not applicable", but that's handled via `tls = null;`.

```suggestion
          type = with types; (listOf (submodule {
```

File: nixos/modules/services/mail/maddy.nix [hunk 14]
```diff
@@ -215,6 +214,65 @@ in {
...
+        extraConfig = mkOption {
+          type = with types; nullOr lines;
+          description = lib.mdDoc ''
+            Arguments for the specific certificate loader. Note that Maddy uses
+            secure defaults for the TLS configuration so there is no need to
+            change anything in most cases.
+            See [upstream manual](https://maddy.email/reference/tls/) for
+            available options.
+          '';
+          default = null;
```
Comment (e1mo) [inline nixos/modules/services/mail/maddy.nix:272]:
```suggestion
          type = types.lines;
          description = lib.mdDoc ''
            Arguments for the specific certificate loader. Note that Maddy uses
            secure defaults for the TLS configuration so there is no need to
            change anything in most cases.
            See [upstream manual](https://maddy.email/reference/tls/) for
            available options.
          '';
          default = "";
```

File: nixos/modules/services/mail/maddy.nix [hunk 15]
```diff
@@ -318,6 +390,17 @@ in {
         $(primary_domain) = ${cfg.primaryDomain}
         $(local_domains) = ${toString cfg.localDomains}
         hostname ${cfg.hostname}
+
+        ${if (cfg.tls.loader == "file") then ''
+          tls file ${concatStringsSep " " (
+            map (x: x.certPath + " " + x.keyPath
+          ) cfg.tls.certificates)} ${optionalString (cfg.tls.extraConfig != null) ''
```
Comment (e1mo) [inline nixos/modules/services/mail/maddy.nix:397]: Or just don't check if it's set and always include the empty parentheses?
```suggestion
          ) cfg.tls.certificates)} ${optionalString (cfg.tls.extraConfig != "") ''
```

**Review by e1mo — CHANGES_REQUESTED** [review]

File: nixos/modules/services/mail/maddy.nix (hunk 9 above)
Comment (onny) [reply nixos/modules/services/mail/maddy.nix]: I updated it and now you'll get the assertion error message with tls enabled if `certificates` is null or empty and if keyPath or certPath is missing in the attribute set. Otherwise the error message might be too cryptic

**Comment by onny** [comment]
Thank you @e1mo for checking it. I simplified the types and assertion according to your recommendations. The assert message works if the `certificates` list is empty. If not both keyPath and certPath are specified, an evaluation error occurs
```
tls = {
  loader = "file";
  certificates = [
    {
      keyPath = "${certs.${domain}.key}";
      certPath = "${certs.${domain}.cert}";
    }
    {
      keyPath = "${certs.${domain}.key}";
    }
  ];
};
```
```
error: The option `services.maddy.tls.certificates."[definition 1-entry 2]".certPath' is used but not defined.
(use '--show-trace' to show detailed location information)
```
If this behavior is fine, I'm okay with it :)

**Comment by e1mo** [comment]
Yepp, that's the behavior I was going/looking for. I think the very slightly less obvious error message is a worthwhile trade-off against a more concise usage of types and effort when it comes to validating arguments. If tests pass, I'm fine happy (just waiting with approval until that's the case).
